pkg/stub/handler.go

```diff
@@ -508,12 +532,10 @@ func (h *Handler) RemoveFinalizer(srcfg *v1alpha1.SamplesResource) {
 }
 
 func (h *Handler) NeedsFinalizing(srcfg *v1alpha1.SamplesResource) bool {
-	h.mutex.Lock()
-	defer h.mutex.Unlock()
-	if h.deleteInProgress {
+	if srcfg.ConditionFalse(v1alpha1.SamplesExist) {
```

Comment:
samplesexist is set to true as soon as we start creating samples? or after all samples are created?

Reply:
I think it's a followup but i think there's still a window here either way.

The conditions you'd need to avoid a race are:

1) SampleCreationInProgress (we are iterating the imagestreams/templates and invoking create on each)
2) SamplesExist (we are done iterating them all, they are created now)

They should never both be true at the same time, and we should not attempt to finalize when (1) is true and (2) is not.

Reply:
The samplesresource with the samples exists condition set to true is persisted to etcd after the samples are created.
The flow either on startup, or after deletion, is:
- the create default logic creates the samplesresource with exists==false, calling the sdk to persist to etcd
- the watch event resulting from that creation comes in
- after the spec validation, mgmt state processing, skip filters creation, etc. we read the samples content from disk and upsert the samples to the openshift namespace
- if errors occur, exists become unknown, persist that to etc,
- is samples upsert successful, exists becomes true and in progress becomes true, and the sdk is called to update ... if conflict error, return error for sdk to relist/retry
- end of initial watch event
- the next watch event for the samples resource, as a result of setting exists,inprogress to true, comes in ... it is effectively ignored since it had no config changes
- the watch events for imagestreams and templates start coming in as soon as we create the samples ... they read in the sdk and will update if  the imagestream is complete (i.e. spec tag generations match status tag generations ... and they update the in progress condition accordingly ... if conflict, return error for retry

I don't agrees that exist==true and inprogress == true cannot occur at the same time

Reply:
I was suggesting a new condition, so you'd have:

CreationInProgress (we are iterating the samples for creation)
SamplesExist (we have created everything we intend to create)
ChangesInProgress (we are waiting for imagestreams to import)

So SamplesExist and ChangesInProgress(ImportInProgress) could both be true.

But CreationInProgress and SamplesExist should be set exclusively (ie if at some point in the future we discover a sample is missing that we need to create, we'd set samplesexist back to false and creationinprogress back to true)

Reply:
OK ... after our discussion and looking at the code, I realized that coupling the inprogress=true change and samplesexists=true change in the same samplesresource attempt like the code currently does is incorrect.

Not the first time I've learned the lesson and then had to re-learn it.

I need to set inprogress to true first, update the object, and then start creating the samples on the next watch event when inprogress is  true.

This essentially what @bparees referred to, but without creating the new condition.

Working on changes now ...